client/core/core.go

```diff
@@ -10498,3 +10498,75 @@ func (c *Core) SendShielded(appPW []byte, assetID uint32, toAddr string, amt uin
 
 	return coinID, nil
 }
+
+// stakingWallet fetches the staking wallet and returns its asset.TicketBuyer
+// interface. Errors if no wallet is currently loaded. Used for ticket
+// purchasing.
+func (c *Core) stakingWallet(assetID uint32) (*xcWallet, asset.TicketBuyer, error) {
+	wallet, exists := c.wallet(assetID)
+	if !exists {
+		return nil, nil, newError(missingWalletErr, "no configured wallet found for %s", unbip(assetID))
+	}
+	ticketBuyer, is := wallet.Wallet.(asset.TicketBuyer)
+	if !is {
+		return nil, nil, fmt.Errorf("%s wallet is not a TicketBuyer", unbip(assetID))
+	}
+	return wallet, ticketBuyer, nil
+}
+
+// StakeStatus returns current staking statuses such as currently owned
+// tickets, ticket price, and current voting preferences. Used for
+// ticket purchasing.
+func (c *Core) StakeStatus(assetID uint32) (*asset.TicketStakingStatus, error) {
+	_, tb, err := c.stakingWallet(assetID)
+	if err != nil {
+		return nil, err
+	}
+	return tb.StakeStatus()
+}
+
+// SetVSP sets the VSP provider. Used for ticket purchasing.
+func (c *Core) SetVSP(assetID uint32, addr string) error {
+	_, tb, err := c.stakingWallet(assetID)
+	if err != nil {
+		return err
+	}
+	return tb.SetVSP(addr)
+}
+
+// PurchaseTickets purchases n tickets. Returns the purchased ticket hashes if
+// successful. Used for ticket purchasing.
+func (c *Core) PurchaseTickets(assetID uint32, pw []byte, n int) ([]string, error) {
+	wallet, tb, err := c.stakingWallet(assetID)
+	if err != nil {
+		return nil, err
+	}
+	crypter, err := c.encryptionKey(pw)
+	if err != nil {
+		return nil, fmt.Errorf("password error: %w", err)
+	}
+	defer crypter.Close()
+	err = c.connectAndUnlock(crypter, wallet)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Send tickets bought notification.
+	//subject, details := c.formatDetails(TopicSendSuccess, sentValue, unbip(assetID), address, coin)
+	//c.notify(newSendNote(TopicSendSuccess, subject, details, db.Success))
+
+	c.updateAssetBalance(assetID)
+	return tb.PurchaseTickets(n)
```

Comment:
Should the balance updating happen _after_ the ticket purchase, if successful? Same with the ntfn TODO a few lines above.

Reply:
Oh yeah, after does make more sense, hmm

Reply:
fixed